### PR TITLE
Added better error message for shared borrow treated as unique for purposes of lifetimes

### PR DESCRIPTION
**Pull request by Polkaverse**
Part of Issue #76630

r? @jyn514

### CODE REVIEW

**Comment by jyn514** [comment]
I only see changes to stderr, not to the compiler itself. You need to find where `mutable borrow starts here` is emitted in the compiler and change that (it will be somewhere in `compiler/`).

Anyway, I'm not the right reviewer for this.

r? @estebank

File: compiler/rustc_mir/src/util/borrowck_errors.rs [hunk 1]
```diff
@@ -68,8 +68,8 @@ impl<'cx, 'tcx> crate::borrow_check::MirBorrowckCtxt<'cx, 'tcx> {
             err.span_label(
                 new_loan_span,
                 format!(
-                    "mutable borrow starts here in previous \
-                     iteration of loop{}",
+                    "mutable already borrowed in the previous \
+                     iteration of the loop{}",
                     opt_via
```
Comment (estebank) [inline compiler/rustc_mir/src/util/borrowck_errors.rs:73]: You probably want to use `desc` here:
```suggestion
                    "{} was borrowed mutably in the previous iteration of the loop{}",
                    desc,
                    opt_via,
```

**Review by estebank — COMMENTED** [review]
You need to run `./x.py test src/test/ui --bless --stage 1` to update the `stderr` files. This will take a while the first time.

**Comment by PankajChaudhary5** [comment]
@estebank  please review.

**Comment by PankajChaudhary5** [comment]
@estebank  please review.

File: compiler/rustc_mir/src/util/borrowck_errors.rs (hunk 1 above)
Comment (camelid) [reply compiler/rustc_mir/src/util/borrowck_errors.rs]: @PankajChaudhary5 Can you please address this comment?

Comment (jyn514) [reply compiler/rustc_mir/src/util/borrowck_errors.rs]: @camelid they did, they just didn't hit the 'resolve' button.

Comment (camelid) [reply compiler/rustc_mir/src/util/borrowck_errors.rs]: @jyn514 They did not address changing the message. I find the current message quite confusing:

> `mutable {}{} already borrowed in the previous iteration of the loop{}`

which renders as:

> `mutable `x` already borrowed in the previous iteration of the loop`

**Comment by PankajChaudhary5** [comment]
Please review @camelid @jyn514

**Comment by jyn514** [comment]
Hmm, 'you have already borrowed' is not really the norm for error messages, normally they're third person. Other than that it looks good though :)

**Comment by PankajChaudhary5** [comment]
@jyn514  is this good?
 `mutable variable `x` is already borrowed in the previous iteration of the loop`

**Comment by jyn514** [comment]
Yup, that looks great!

File: compiler/rustc_mir/src/util/borrowck_errors.rs [hunk 2]
```diff
@@ -68,9 +68,10 @@ impl<'cx, 'tcx> crate::borrow_check::MirBorrowckCtxt<'cx, 'tcx> {
             err.span_label(
                 new_loan_span,
                 format!(
-                    "mutable borrow starts here in previous \
-                     iteration of loop{}",
-                    opt_via
+                    "mutable variable {}{} is already borrowed in the previous iteration of the loop{}",
```
Comment (camelid) [inline compiler/rustc_mir/src/util/borrowck_errors.rs:71]: I find this error confusing: the issue is not that _x_ is mutable, rather that it was mutably _borrowed_. How about something like this?

```suggestion                    "{}{} was already mutably borrowed in the previous iteration of the loop{}",
```

**Review by camelid — CHANGES_REQUESTED** [review]

**Comment by PankajChaudhary5** [comment]
@camelid @jyn514

File: compiler/rustc_mir/src/util/borrowck_errors.rs [hunk 3]
```diff
@@ -68,9 +68,10 @@ impl<'cx, 'tcx> crate::borrow_check::MirBorrowckCtxt<'cx, 'tcx> {
             err.span_label(
                 new_loan_span,
                 format!(
-                    "mutable borrow starts here in previous \
-                     iteration of loop{}",
-                    opt_via
+                    "{}{} was already mutably borrowed in the previous iteration of the loop{}",
```
Comment (camelid) [inline compiler/rustc_mir/src/util/borrowck_errors.rs:71]: I realized that the error I suggested was not ideal. Would you be open to this instead? Sorry I keep asking you to change it! I just want to make sure the error is as clear as possible :)

```suggestion
                    "{}{} was mutably borrowed here in the previous iteration of the loop{}",
```

**Review by camelid — APPROVED** [review]
Looks great, thanks!

**Comment by PankajChaudhary5** [comment]
@camelid can you merge it?

**Comment by camelid** [comment]
No, I think we'll have to wait for @estebank or someone else to merge it (estebank tends to be very busy).

**Comment by PankajChaudhary5** [comment]
@GuillaumeGomez  can you please take a look at it

**Comment by camelid** [comment]
I don't think Guillaume is the right person to review this.

r? @davidtwco (feel free to re-assign)

**Review by GuillaumeGomez — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
It's indeed been a long time since I last worked on error messages so I'll let someone else r+ it (even though I approved it).

**Review by davidtwco — APPROVED** [review]

**Pull request merged**